cip-ledger-state.md

```diff
@@ -0,0 +1,111 @@
+---
+CIP: ?
+Title: Extended Local Chain Sync Protocol
+Authors: Erik de Castro Lopo <erikd@mega-nerd.com>
+Discussions-To: Erik de Castro Lopo <erikd@mega-nerd.com>
+Comments-Summary: Extend Local Chain Sync Protocol
+Comments-URI:
+Status: Draft
+Type: ?
+Created: 2022-11-15
+License: CC-BY-4.0
+---
+## Abstract
+
+Modify the `cardano-node` (and underlying code) to provide an extended version of the existing
+local chain sync protocol.
+
+## Motivation
+
+Applications that provide insight into the Cardano block chain (like db-sync, exporters, Kupo, and
+smart contracts reacting to events) often need access to the current state of the Cardano ledger
+(stake distribution, reward and wallet balances etc). This information is known to the node, partly
+on the block chains and partly in what is referred to as ledger state (described more fully below).
+Extracting block chain data is relatively easy, but ledger state data is not. Currently these
+applications have to recreate and maintain ledger state themselves based on the block information
+they stream from the node over the local chain sync protocol. Recreation and maintenance of the
+ledger state is not only complex and a source of bugs but more importantly requires significant
+resources, specifically RAM. Ledger state in memory currently consumes 10 Gig of RAM and that is
+growing. In situations where the node and any such application run on the same machine, the machine
+ends up with twice the resource usage. The following proposal hopes to reduce resource usage and
+complexity for chain following applications like db-sync.
+
+# Current Situation
+
+Currently there is a local chain sync protocol which is really just the peer-to-peer protocol
+using a local domain socket rather than the TCP/IP socket normally used for P2P transport.
+
+The data transported over this local chain sync protocol is limited to block chain data. However, a
+Cardano node also maintains ledger state which includes:
+
+* The current UTxO state.
+* Current amount of ADA delegated to each stake pool.
+* Which stake address is currently delegated to each pool.
+* Rewards account balances for each stake address.
+* Current protocol parameters.
+
+The first of these ledger state components is by far the largest component and is probably not
+needed outside the node (and definitely not needed by db-sync. However the others are needed and
+stored by `cardano-db-sync` which gets these data sets by maintaining its own copy of ledger state
+and periodically extracting the parts required.
+
+This means that when `node` and `db-sync` are run on the same machine (which is the recommended
+configuration) that machine has two basically identical copies of ledger state. Ledger state is a
+*HUGE* data structure and the mainnet version currently consumes over 10 Gigabytes of memory.
+Furthermore, maintaining ledger state duplicates code functionality that is in `ouroboros-consensus`
```

Comment:
Can't this idea be extended even more to solve the duplicate RAM problem? What if db-sync maintained its own ChainDB, connected to peers itself, ran ouroboros etc? That way it doesn't need a separate trusted node process, there is no replication, no need for protocol extension, deployment is greatly simplified etc. At the same time this adds no complexity to the node that @dcoutts understandably wants to avoid, as this complexity will live in the db-sync repository.

Reply:
Another alternative -- and straightforward -- solution would be to also simply "dump" information required to the calculation of implicit stuff (such as rewards) into some file as the node processes them. And either, document the format of that file or provide a thin api-layer to query its content. 

Having just that would solve many problems down the line; client applications will almost always be connected to a local node and that node will have to do this work anyway. So instead of duplicating that work; simply record the information when its done and gives way to access that information later on. It could be a mere .csv or .json file recording the stake distribution on each epoch transition.

Reply:
Dumping to files seems to be the most sensible idea.
Another option is to keep the information in memory. This would require some backpressure (eg using a TBQueue) to make sure memory usage doesn't explode. When the queue gets full the syncing/chainselection in node will stop. That way we could build a data node, which processes the ledger events in its own pace.

This still requires change in the ChainDB API.

Reply:
> Make the full node's logic available as a library instead of only as an executable

Doesn't this still mean there are two copies of ledger state?

Reply:
Not if db-sync IS the node, eg. it embeds the node's logic, is connected to the network, runs the consensus and ledger's rules, etc.

Reply:
Oh, ok, so you would not need to run a node on a machine running `db-sync` at all. Not sure how this compares to mu proposal in terms of amount of work or complexity.

Reply:
It's probably (much) more work so I'd not consider this as an alternative for this particular CIP

Reply:
Funny in the last company I worked for, everything was written as a library first (makes testing much easier) and then thin wrapper was added to make it into an executable.

Reply:
> Then db-sync or any client could use that "ghost node" to replay blocks from any point in time, generate events, query ledger state and what not. I am even surprised such a tool does not already exist: Is this not somewhat akin to what db-analyser do?

Is this similar to what Santiago from TxPipe proposed a while back called Dolos, a Data Node?

Reply:
I don't know, I wasn't there a while back ;) But I guess this idea is pretty much obvious and not novel so I am pretty sure other people have thought of iti.